e2e/internal/e2e/singularitycmd.go

```diff
@@ -0,0 +1,354 @@
+	return func(t *testing.T, c *expect.Console) {
+		if o, err := c.Expectf(format, args...); err != nil {
+			expected := fmt.Sprintf(format, args...)
+			t.Logf("\nConsole output: %s\nExpected output: %s", o, expected)
+			t.Errorf("error while reading from console: %s", err)
+		}
+	}
+}
+
+// ConsoleExpect reads from console until the provided string is read or
+// an error occurs.
+func ConsoleExpect(s string) SingularityConsoleOp {
+	return func(t *testing.T, c *expect.Console) {
+		if o, err := c.ExpectString(s); err != nil {
+			t.Logf("\nConsole output: %s\nExpected output: %s", o, s)
+			t.Errorf("error while reading from console: %s", err)
+		}
+	}
+}
+
+// ConsoleSend writes string to console.
+func ConsoleSend(s string) SingularityConsoleOp {
+	return func(t *testing.T, c *expect.Console) {
+		if _, err := c.Send(s); err != nil {
+			t.Errorf("error while writing string to console: %s", err)
+		}
+	}
+}
+
+// ConsoleSendLine writes string to console with a trailing newline.
+func ConsoleSendLine(s string) SingularityConsoleOp {
+	return func(t *testing.T, c *expect.Console) {
+		if _, err := c.SendLine(s); err != nil {
+			t.Errorf("error while writing string to console: %s", err)
+		}
+	}
+}
+
+// SingularityCmdOp is a function type passed to RunCommand
+// used to define the test execution context.
+type SingularityCmdOp func(*singularityCmd)
+
+// singularityCmd defines a Singularity command execution test.
+type singularityCmd struct {
+	args       []string
+	envs       []string
+	dir        string
+	privileged bool
+	stdin      io.Reader
+	preFn      func(*testing.T)
+	postFn     func(*testing.T)
+	consoleFn  SingularityCmdOp
+	console    *expect.Console
+	resultFn   SingularityCmdOp
+	result     *SingularityCmdResult
+	waitErr    error
+	t          *testing.T
+}
+
+// WithCommand sets singularity command to execute.
+func WithCommand(command string) SingularityCmdOp {
+	return func(s *singularityCmd) {
+		cmd := strings.Split(command, " ")
+		s.args = append(cmd, s.args...)
+	}
+}
+
+// WithArgs sets singularity command arguments.
+func WithArgs(args ...string) SingularityCmdOp {
+	return func(s *singularityCmd) {
+		if len(args) > 0 {
+			s.args = append(s.args, args...)
+		}
+	}
+}
+
+// WithEnv sets environment variables to use while running
+// singularity command.
+func WithEnv(envs []string) SingularityCmdOp {
+	return func(s *singularityCmd) {
+		if len(envs) > 0 {
+			s.envs = append(s.envs, envs...)
+		}
+	}
+}
+
+// WithDir sets current working directory for execution.
+func WithDir(dir string) SingularityCmdOp {
+	return func(s *singularityCmd) {
+		if dir != "" {
+			s.dir = dir
+		}
+	}
+}
+
+// WithPrivileges sets if singularity command must be
+// executed with privileges or not. PreRun, InRun, PostRun
+// are also executed with privileges.
+func WithPrivileges(privileged bool) SingularityCmdOp {
+	return func(s *singularityCmd) {
+		cacheDirEnv := fmt.Sprintf("SINGULARITY_CACHEDIR=%s", cacheDirPriv)
```

Comment:
If this does not create a dependency cycle, maybe we should use `cache.DirEnv` instead of hardcoding `SINGULARITY_CACHEDIR`